Add more color to query language docs (#154)
ShiftLeftSecurity/joern: docs/content/querying/_index.md

+++
Title="Query Language"
weight=5
alwaysopen=true
+++

Joern/Ocular provides a domain specific language (DSL) for code
analysis based on [Scala](https://www.scala-lang.org/). It is an
internal DSL, that is, it is a language defined via classes and
methods as opposed to a formal grammar. The great advantage of this
approach is that all features of the host language are automatically
inherited. Joern queries are lazily evaluated Scala expressions, and
you can make use of the complete Scala programming language in Joern
scripts. In this section, we look into the basic concepts of our query
language and illustrate them by example.

In the following examples, we assume that a code property graph for
the *tarpitc* program is loaded. For details on creating and loading
CPGs, please check out [Importing Code](../importing) and [The Joern
Shell](../shell).

# Nodes, Labels, and Expressions

> To get a first idea of the data which the language makes accessible,

Reviewer: ```suggestion
To begin exploring the data that the language exposes,
```

> let us begin querying for all nodes of the code property graph via

Reviewer: ```suggestion
let's create a query that selects all nodes in the code property graph with
```

`cpg.all`:

```bash
cpg.all
res8: NodeSteps[StoredNode] = io.shiftleft.semanticcpg.language.NodeSteps@4cce421e
```

Notice that the result is not a list of nodes but rather a variable type
`NodeSteps[StoredNode]`. You can think of this as an *expression*

> which - only upon evaluation - yields all nodes [1]. The advantage of

Reviewer: ```suggestion
which only upon evaluation will yield all nodes [1]. The advantage of
```

this two stage approach is that we can already perform calculations on
expressions prior to costly evaluation. In particular, we can use the
type system to perform validity checks and query optimization,

> features that are vital as we compose expression to form complex

Reviewer: ```suggestion
features that are vital as we compose expressions to form complex
```

list and sorted. Note that both `toList` and `sorted` are API methods
of the host language Scala.

[1] The term `Step` is inherited from the graph traversal language
[Gremlin](http://tinkerpop.apache.org/docs/current/reference/#graph-traversal-steps),
and more specifically
[gremlin-scala](https://github.com/mpollmeier/gremlin-scala) on which
our DSL is based. In Gremlin, `Steps` describe walks in a graph, which
are combined via function composition to yield *traversals*. You can
think of traversals as expressions that describe sets of nodes.

## Expression Types and Starters

> For each node label, a corresponding *expression type* is defined in the

Reviewer: I'm not 100% sure if this change makes sense, but just calling it a `type` seems more intuitive to me.

```suggestion
For each node label, a corresponding *type* is defined in the
```

query language, e.g., `Method` for nodes with the label `METHOD`, and
`Local` for nodes with the label `LOCAL`. For each label, we also
define a *starter* that represents all nodes with that label. For
example,

```bash
cpg.method
res1: types.structure.Method = io.shiftleft.semanticcpg.language.types.structure.Method@b889cb6
```
returns an expression that yields all methods on evaluation.

You can get a complete list of starters by typing `cpg.`, and
pressing the Tab key:

```bash
joern> cpg.
all             file            local           methodReturn    types
argument        graph           member          namespace
call            id              metaData        namespaceBlock
close           identifier      method          parameter
comment         literal         methodRef       typeDecl

```

It is generally advisable to use specific starters as opposed to the
generic `cpg.all`, as for the latter, it is unclear at compile time
which types of nodes are processed. Using specific starters such as
`.method` make it possible to use the Scala's type system for

> completion and to detect invalid queries even before evaluation.

Reviewer: ```suggestion
auto-completion and to detect invalid queries before they are evaluated.
```

## Node IDs

For each node, we store an **ID**. The ID is a long integer that
uniquely identifies the node.You should not need to ever retrieve
node IDs within Joern/Ocular scripts, however, they may come in handy
when passing nodes to external tools - especially if these tools plan
to run subsequent queries.

As an example, the following query returns the ids of all file nodes.

```bash
cpg.file.id.l
res11: List[AnyRef] = List(34L, 32L, 30L, 28L, 26L, 24L, 22L, 20L, 18L, 16L, 14L, 12L, 10L, 8L, 6L, 4L)
```

Consider that - upon receiving this result, an external tool would
like to invoke a follow-up query on the first file node. To achieve
this, it can issue the query

```bash
cpg.id(34).l
res7: List[Nothing] = List(
  File(id -> 10L, name -> "/home/tmp/shiftleft/joern/tarpit-c/tarpitc/buffer_underwrite.c", order -> null)
)
```

Note that the type of the nodes returned by `cpg.id(34)` is unknown at
compile time, and therefore, the returned type for `cpg.id(34).l`is

> `List[Nothing]`. In practice, the external tool will usually know the

Reviewer: Not related to the content, but a returning a `List[Nothing]` from this seems super strange as it is a bottom type. I would expect `Any` or `AnyRef`.

node type from the previous query, e.g., in our example, we know that

> the node with the ID `34` is a file node. An explicit cast can be

Reviewer: ```suggestion
the node with the ID `34` is a `File` node. An explicit cast can be
```
